enhancement(http_request): add redaction for sensitive headers

benjamin-awd opened the pull request:
## Summary
This PR adds redaction for sensitive headers within the http_request function.

Currently, sensitive information can be leaked at runtime when the function fails: 
```
function call error for "http_request" at (4:263): HTTP request failed: Request failed after 3 
  retries (url: http://127.0.0.1:1/test, method: GET, headers: { "Authorization": "Bearer 
  secret_12345", "Content-Type": "application/json", "X-Api-Key": "key-67890" })
```

With the changes within this PR:
```
  function call error for "http_request" at (4:263): HTTP request failed: Request failed after 3 
  retries (url: http://127.0.0.1:1/test, method: GET, headers: { "Authorization": "***", 
  "Content-Type": "application/json", "X-Api-Key": "***" })
```

## Change Type

- [x] Bug fix
- [ ] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

Unit tests

Run VRL binary locally:
./target/debug/vrl

```
$ . = http_request!("http://127.0.0.1:1/test",
    method: "GET",
    headers: {
        "Authorization": "Bearer secret_12345",
        "X-Api-Key": "key-67890",
        "Content-Type": "application/json"
    }
)
function call error for "http_request" at (4:211): HTTP request failed: 
Request failed after 3 retries (url: http://127.0.0.1:1/test, method: GET, headers: 
{ "Authorization": "***", "Content-Type": "application/json", "X-Api-Key": "***" })
```


## Does this PR include user facing changes?

- [ ] Yes. Please add a changelog fragment based on
  our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [x] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [ ] Our [CONTRIBUTING.md](https://github.com/vectordotdev/vrl/blob/main/CONTRIBUTING.md) is a good starting place.
- [ ] If this PR introduces changes to [LICENSE-3rdparty.csv](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv), please
  run `dd-rust-license-tool write` and commit the changes. More details [here](https://crates.io/crates/dd-rust-license-tool).
- [ ] For new VRL functions, please also create a sibling PR in Vector to document the new function.

<!-- Examples for the above:
  PR adding new VRL function: https://github.com/vectordotdev/vrl/pull/993
  PR adding documentation: https://github.com/vectordotdev/vector/pull/21142
  
  We are working towards improving this workflow.
-->

## References

<!-- Please list any issues closed by this PR. -->

<!--
- Closes: <issue link>
-->

<!-- Any other issues or PRs relevant to this PR? Feel free to list them here. -->

benjamin-awd committed:
enhancement(http_request): add redaction for sensitive headers

thomasqueirozb reviewed (commented):
Hey @benjamin-awd, this is a nice to have. Do you think we can also make this on-by-default and have a parameter that controls whether this happens or not?

benjamin-awd committed:
refactor(http_request): lazily compute redacted headers on error only

benjamin-awd committed:
enhancement(http_request): add redact_headers parameter to control header redaction

Defaults to true so sensitive headers are redacted in error messages.
Can be set to false for debugging purposes.

benjamin-awd committed:
chore(http_request): remove unused `use tokio` import

benjamin-awd committed:
refactor(http_request): use `then_some` in make_proxies

benjamin-awd committed:
refactor(http_request): hoist clippy::similar_names allow to module level

pront commented on changelog.d/1562_http_request_redaction.enhancement.md:
```diff
@@ -0,0 +1 @@
+The `http_request` function now redacts sensitive header values in error messages to prevent credential leakage in logs. Headers such as `Authorization`, `Cookie`, `X-Api-Key`, and any header containing "token", "secret", or "password" will now display as `***` in error output, while non-sensitive headers remain visible for debugging purposes.
```
If I recall correctly we do not publicly advertise this function so we do not need a changelog here (if this is a good idea is another discussion).

benjamin-awd replied:
Ah yes I had forgotten about that, removed changeling in [2e4315b](https://github.com/vectordotdev/vrl/pull/1562/commits/2e4315b52dddb397d4e9a73c6c35e2097c0baf99)

benjamin-awd committed:
chore: remove changelog fragment

benjamin-awd committed:
Merge upstream/main into redact-http-secret

Resolve conflict in parameters() by adopting main's PARAMETERS static
with the redact_headers parameter added. Fix test struct fields to use
Option wrappers matching main's HttpRequestFn changes.

thomasqueirozb committed:
Merge remote-tracking branch 'origin/main' into redact-http-secret

thomasqueirozb committed:
Update description

thomasqueirozb committed:
Use DEFAULT_REDACT_HEADERS in resolver

thomasqueirozb reviewed (approved):
Fixed merge conflicts and did some touching up to make this conform to existing/new code patterns in the codebase (see last few commits).

Thanks for your contribution!

thomasqueirozb committed:
Revert "chore: remove changelog fragment"

This reverts commit 2e4315b52dddb397d4e9a73c6c35e2097c0baf99.

thomasqueirozb commented:
(also re-added the changelog - this function will be documented before the next release)

benjamin-awd committed:
Merge branch 'main' into redact-http-secret
